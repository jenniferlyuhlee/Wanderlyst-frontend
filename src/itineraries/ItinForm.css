.Form-tag-check{
    margin: 0 0 1rem 0;
    padding: 0;
}

.Form-tag-check .form-check {
    width: 20%; /* 5 columns for screens larger than 1000px */
    display: inline-block;
}
  
  
  @media (max-width: 900px) {
    .Form-tag-check .form-check {
      width: 50%; /* 2 columns for screens smaller than 1000px */
      display: inline-block;
    }
  }