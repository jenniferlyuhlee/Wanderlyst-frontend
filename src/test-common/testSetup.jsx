import React, { useState } from "react";
import UserContext from "../auth-user/UserContext";
import WanderlystApi from "../utils/api";
import '@testing-library/jest-dom';

const demoUser = {
  username: "testadmin", 
  firstName: "testfirst",
  lastName: "testlast",
  location: "testloc",
  bio: "testbio",
  profilePic: null,
  isAdmin: true,
};

const demoTags = [
  {id: 1, name: 'Foodie', description: 'Savor the world one bite at a time. Dive into a journey packed with delicious local cuisines, hidden culinary gems, and must-try dishes that will leave your taste buds tingling.'},
  {id: 2, name: 'Explorer', description: 'Step off the beaten path and uncover the unknown. From ancient ruins to vibrant cityscapes, let your curiosity guide you through unforgettable experiences.'}
]

const UserProvider = ({ children, 
                        currUser = demoUser, 
                        setCurrUser, 
                        tags = demoTags,
                        logout,
                        hasLikedItin,
                        toggleLike }) => {
  const [user, setUser] = useState(currUser);

  return (
    <UserContext.Provider value={{ currUser: user, 
                                  setCurrUser: setUser, 
                                  tags, logout, 
                                  hasLikedItin, toggleLike}}>
      {children}
    </UserContext.Provider>
  );
};

export { UserProvider };