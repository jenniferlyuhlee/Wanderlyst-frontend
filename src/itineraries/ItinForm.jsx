import React, { useState, useContext} from "react";
import { useNavigate } from "react-router-dom";
import WanderlystApi from "../utils/api";
import cleanData from "../helpers/cleanData";
import TagsSelect from "./TagsSelect";
import PlaceForm from "./PlaceForm";
import Alert from "../shared/Alert";
import "./ItinForm.css"

function ItinForm(){
    const initialPlaceState = {
        name: "",
        address: "",
        description: "",
        image: ""
    }
    const initialItinState = {
        title: "",
        city: "",
        country: "",
        duration: "",
        description: "",
        tags: [],
        places: [initialPlaceState]
    }

    // states: formData, formErrors, saveConfirmed
    const [formData, setFormData] = useState(initialItinState);
    const [formErrors, setFormErrors] = useState(null);

    const navigate = useNavigate();

    // general callback to update targeted form data for basic fields
    function handleChange (e) {
        const { name, value } = e.target;
        setFormData(data => ({...data, [name]: value }));
    }

    // Updates tags data 
    function handleTagChange(e){
        const { value, checked } = e.target;
    setFormData(data => {
        if (checked) {
            // Add tag only if it's not already in the array
            return {
                ...data,
                tags: data.tags.includes(value) ? data.tags : [...data.tags, value]
            };
        } else {
            // Remove tag if unchecked
            return {
                ...data,
                tags: data.tags.filter(t => t !== value)
            };
        }
    });
    }

    // Updates place form data based on index of place
    function handlePlaceChange(index, e){
        const { name, value } = e.target;
        const updatedPlaces = [...formData.places];
        updatedPlaces[index] = {...updatedPlaces[index], [name]: value}
        setFormData(data => ({...data, places: updatedPlaces}));
    }

    // handler for adding new place form when button clicked
    function addPlace(){
        setFormData(data => ({
            ...data,
            places: [...data.places, initialPlaceState]
        }));
    }

    // handler for removing new place form when button clicked
    function removePlace(idx){
        setFormData(data => ({
            ...data,
            places: data.places.slice(0, -1)
        }));
    }

    async function handleSubmit(e){
        e.preventDefault();
        try{
            const itineraryData = {
                ...formData,
                // automatically add sequence to data without user specifying
                places: formData.places.map((place, i) => ({
                        ...cleanData(place),
                        seq: i + 1
                    })
                )
            };
            console.log(itineraryData);
    
            const itinerary = await WanderlystApi.createItin(itineraryData);
            navigate(`itineraries/${itinerary.id}`)
        }
        catch(err){
            setFormErrors(err)
            return;
        }
    }

    return(
         <div className = "container mb-5 body-cont">
            <div className = "card px-5">
                <h2 className = "text-center">
                    {formData.title || "My New Itinerary"}
                </h2>
                <form onSubmit = {handleSubmit} className = "row m-1">
                    <div className = "form-floating col-9 mb-2 p-0">
                        <input id = "title"
                            className = "form-control"
                            name = "title"
                            placeholder = "firstName"
                            value = {formData.title}
                            onChange = {handleChange}
                        />
                        <label htmlFor = "title">Title</label>
                    </div>
                    <div className ="col-3 form-floating p-0">
                    <input id = "duration"
                            name = "duration"
                            className = "form-control"
                            type = "number"
                            min = "1"
                            placeholder = "duration"
                            value = { formData.duration }
                            onChange = {handleChange} />
                    <label htmlFor = "duration">Days</label>
                </div>
                    <div className = "form-floating col-6 mb-2 p-0">
                        <input id = "city"
                            className = "form-control"
                            name = "city"
                            placeholder = "city"
                            value = {formData.city}
                            onChange = {handleChange}
                        />
                        <label htmlFor = "city">City</label>
                    </div>
                    <div className = "form-floating col-6 mb-2 p-0">
                        <input id = "country"
                            className = "form-control"
                            name = "country"
                            placeholder = "country"
                            value = {formData.country}
                            onChange = {handleChange}
                        />
                        <label htmlFor = "country">Country</label>
                    </div>
                    <div className = "form-group col-12 mb-2 p-0">
                        <textarea id = "description"
                            className = "form-control"
                            name = "description"
                            rows = "4"
                            placeholder = "Share a description of the overall itinerary! This will display on the itinerary preview card."
                            value = {formData.description}
                            onChange = {handleChange}
                        />
                    </div>

                    {/* TagsSelect Form component */}
                    <TagsSelect selectedTags={formData.tags} handleTagChange={handleTagChange}/>
                    
                    {/* Render PlaceForm for each place added in Itinerary.places */}
                    {formData.places.map((place, i) => (
                        <PlaceForm key={i} 
                                   index={i} 
                                   placeData={place} 
                                   handlePlaceChange={handlePlaceChange}
                        />
                    ))}

                    <div className = "p-0 text-center">
                        <button type="button" className = "btn p-0 mb-2 mx-2 secondary" 
                        onClick={addPlace}>
                            <i className="bi bi-plus-circle-fill"></i>
                        </button>
                    {formData.places.length > 1 ? 
                        <button type="button" className = "btn p-0 mb-2 mx-2 secondary" 
                        onClick={removePlace}>
                            <i className="bi bi-dash-circle-fill"></i>
                        </button>
                        :
                        null
                    }
                    </div>

                    {formErrors?
                        <Alert type = "danger" messages = {formErrors}/>
                        :
                        null
                    }

                    <button className = "btn rounded-pill btn-lg btn-primary">
                        Share
                    </button>
                </form>
            </div>
        </div>
    )
}

export default ItinForm;