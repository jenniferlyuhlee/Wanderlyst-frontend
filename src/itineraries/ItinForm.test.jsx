import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { UserProvider } from "../test-common/testSetup";
import { MemoryRouter} from "react-router-dom";
import WanderlystApi from "../utils/api";
import ItinForm from "./ItinForm";
import { beforeEach, afterEach, expect, describe, test} from "vitest";

describe("ItinForm component", () => {
    // Mock API method to prevent actual submission and redirect
    beforeEach(() => {
        vi.spyOn(WanderlystApi, "createItin").mockResolvedValue(false);
    });

    afterEach(() => {
        vi.restoreAllMocks();
    });

    test("renders without crashing", () => {
        render(
            <MemoryRouter>
                <UserProvider>
                    <ItinForm />
                </UserProvider>
            </MemoryRouter>)
    });
    test("matches snapshot", () => {
        const {asFragment} = render(
            <MemoryRouter>
                <UserProvider>
                    <ItinForm />
                </UserProvider>
            </MemoryRouter>);
        expect(asFragment()).toMatchSnapshot();
    });

    test("handleChange being called when form fields change", async () => {
        const {getByLabelText, getByText} = render(
            <MemoryRouter>
                <UserProvider>
                    <ItinForm />
                </UserProvider>
            </MemoryRouter>);
        
        // check that regular form fields being controlled
        const titleInput = getByLabelText("Title");
        
        // fire change event
        fireEvent.change(titleInput, {target: {value: "Test Itin"}});
        await waitFor(()=> {
            expect(titleInput.value).toBe("Test Itin");
        })
    });

    test("handleTagChange being called when tags selected", async () => {
        const {getByLabelText} = render(
            <MemoryRouter>
                <UserProvider>
                    <ItinForm />
                </UserProvider>
            </MemoryRouter>);
        
        // check that regular form fields being controlled
        const tagInput = getByLabelText("Foodie");
        
        // fire change event
        fireEvent.click(tagInput)
        await waitFor(()=> {
            expect(tagInput.checked).toBe(true);
        });
        fireEvent.click(tagInput)
        await waitFor(()=> {
            expect(tagInput.checked).toBe(false);
        });
    });

    test("handlePlaceChange being called when place form is updated", async () => {
        const {debug, getByLabelText} = render(
            <MemoryRouter>
                <UserProvider>
                    <ItinForm />
                </UserProvider>
            </MemoryRouter>);
        debug();
        // check that regular form fields being controlled
        const placeInput = getByLabelText("Place Name");
        
        // fire change event
        fireEvent.change(placeInput, {target: {value: "Test Place"}});
        await waitFor(()=> {
            expect(placeInput.value).toBe("Test Place");
        })
    });

    test("calls createItin when place form is updated", async () => {
        const { container, getByLabelText, getByText} = render(
            <MemoryRouter>
                <UserProvider>
                    <ItinForm />
                </UserProvider>
            </MemoryRouter>);

        // fill out form fields
        fireEvent.change(getByLabelText("Title"), {target: {value: "test title"}});
        fireEvent.change(getByLabelText("Days"), {target: {value: "1"}});
        fireEvent.change(getByLabelText("City"), {target: {value: "test city"}});
        fireEvent.change(getByLabelText("Country"), {target: {value: "test country"}});
        fireEvent.change(container.querySelector("#description"), {target: {value: "test desc"}});
        fireEvent.click(getByLabelText("Foodie"));
        fireEvent.change(getByLabelText("Place Name"), {target: {value: "test place"}});
        fireEvent.change(getByLabelText("Address"), {target: {value: "test address"}});
        fireEvent.change(container.querySelector("#description1"), {target: {value: "test place desc"}});
        
        // "submit" form
        const btn = getByText("Share");
        fireEvent.click(btn);
        
        // wait for mock func to be called
        await waitFor(() => {
            expect(WanderlystApi.createItin).toHaveBeenCalledWith({
                    city: "test city",
                    country: "test country",
                    description: "test desc",
                    duration: "1",
                    places: [
                            {
                             address: "test address",
                             description: "test place desc",
                             name: "test place",
                             seq: 1,
                            },
                         ],
                    tags:  ["1"],
                    title: "test title",
                }
            );
        });
    });
});