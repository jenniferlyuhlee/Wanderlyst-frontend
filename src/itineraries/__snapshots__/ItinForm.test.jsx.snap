// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ItinForm component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container mb-5 body-cont"
  >
    <div
      class="card px-5"
    >
      <h2
        class="text-center"
      >
        My New Itinerary
      </h2>
      <form
        class="row m-1"
      >
        <div
          class="form-floating col-9 mb-2 p-0"
        >
          <input
            class="form-control"
            id="title"
            name="title"
            placeholder="firstName"
            value=""
          />
          <label
            for="title"
          >
            Title
          </label>
        </div>
        <div
          class="col-3 form-floating p-0"
        >
          <input
            class="form-control"
            id="duration"
            min="1"
            name="duration"
            placeholder="duration"
            type="number"
            value=""
          />
          <label
            for="duration"
          >
            Days
          </label>
        </div>
        <div
          class="form-floating col-6 mb-2 p-0"
        >
          <input
            class="form-control"
            id="city"
            name="city"
            placeholder="city"
            value=""
          />
          <label
            for="city"
          >
            City
          </label>
        </div>
        <div
          class="form-floating col-6 mb-2 p-0"
        >
          <input
            class="form-control"
            id="country"
            name="country"
            placeholder="country"
            value=""
          />
          <label
            for="country"
          >
            Country
          </label>
        </div>
        <div
          class="form-group col-12 mb-2 p-0"
        >
          <textarea
            class="form-control"
            id="description"
            name="description"
            placeholder="Share a description of the overall itinerary! This will display on the itinerary preview card."
            rows="4"
          />
        </div>
        <div
          class="Form-tag-check row mt-2"
        >
          <h5
            class="mb-1 p-0"
          >
            Tags:
          </h5>
          <div
            class="form-check"
          >
            <input
              class="form-check-input col"
              id="Foodie"
              type="checkbox"
              value="1"
            />
            <label
              class="form-check-label"
              for="Foodie"
            >
              Foodie
            </label>
          </div>
          <div
            class="form-check"
          >
            <input
              class="form-check-input col"
              id="Explorer"
              type="checkbox"
              value="2"
            />
            <label
              class="form-check-label"
              for="Explorer"
            >
              Explorer
            </label>
          </div>
        </div>
        <div
          class="place-form my-1 p-0"
        >
          <h5
            class="mb-1 p-0"
          >
            Place 1:
          </h5>
          <div
            class="form-floating mb-2 p-0"
          >
            <input
              class="form-control"
              id="name1"
              name="name"
              placeholder="Place Name"
              required=""
              value=""
            />
            <label
              for="name1"
            >
              Place Name
            </label>
          </div>
          <div
            class="form-floating mb-2 p-0"
          >
            <input
              class="form-control"
              id="address1"
              name="address"
              placeholder="Place Address"
              required=""
              value=""
            />
            <label
              for="address1"
            >
              Address
            </label>
          </div>
          <div
            class="form-group mb-2 p-0"
          >
            <textarea
              class="form-control"
              id="description1"
              name="description"
              placeholder="Share what you did or can do here! Plus any time expectations, highlights, tips, important precautions etc. "
              required=""
              rows="3"
            />
          </div>
          <div
            class="form-floating mb-2 p-0"
          >
            <input
              class="form-control"
              id="image1"
              name="image"
              placeholder="Image"
              value=""
            />
            <label
              for="image1"
            >
              Image URL (optional)
            </label>
          </div>
        </div>
        <div
          class="p-0 text-center"
        >
          <button
            class="btn p-0 mb-2 mx-2 secondary"
            type="button"
          >
            <i
              class="bi bi-plus-circle-fill"
            />
          </button>
        </div>
        <button
          class="btn rounded-pill btn-lg btn-primary"
        >
          Share
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
